src/metabase/query_processor/card.clj

```diff
@@ -150,19 +170,39 @@
 
 (s/defn ^:private validate-card-parameters
   "Unless [[*allow-arbitrary-mbql-parameters*]] is truthy, check to make all supplied `parameters` actually match up
-  with template tags in the query for Card with `card-id`."
-  [card-id :- su/IntGreaterThanZero parameters :- mbql.s/ParameterList]
+  with `parameters` defined on the Card with `card-id`, or template tags on the query."
+  [card-id :- su/IntGreaterThanZero request-parameters :- mbql.s/ParameterList]
   (when-not *allow-arbitrary-mbql-parameters*
-    (let [template-tags (card-template-tag-parameters card-id)]
-      (doseq [request-parameter parameters
-              :let              [parameter-name (infer-parameter-name request-parameter)]]
-        (let [matching-widget-type (or (get template-tags parameter-name)
-                                       (throw (ex-info (tru "Invalid parameter: Card {0} does not have a template tag named {1}."
-                                                            card-id
-                                                            (pr-str parameter-name))
-                                                       {:type               qp.error-type/invalid-parameter
-                                                        :invalid-parameter  request-parameter
-                                                        :allowed-parameters (keys template-tags)})))]
+    (let [card               (api/check-404 (db/select-one [Card :dataset_query :parameters] :id card-id))
+          template-tag-types (card-template-tags card)
```

Comment:
wdyt if we make this more explicit? , sth like `types-on-template-tag` and `types-on-parameter`

Reply:
sure